core/arch/arm/kernel/rpc_io_i2c.c

```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2020 Foundries Ltd <jorge@foundries.io>
+ *
+ */
+#include <kernel/rpc_io_i2c.h>
+#include <kernel/thread.h>
+#include <mm/mobj.h>
+#include <optee_rpc_cmd.h>
+#include <string.h>
+
+#define I2C_BUFFER_LENGTH 512 /* enough for most use cases */
+static struct mobj *mobj;
+
+TEE_Result rpc_io_i2c_transfer(struct rpc_i2c_request *req, size_t *bytes)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct thread_param p[3] = { };
+
+	if (!req)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (req->buffer_len > I2C_BUFFER_LENGTH) {
+		EMSG("buffer too large to transfer");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (!mobj) {
+		mobj = thread_rpc_alloc_kernel_payload(I2C_BUFFER_LENGTH);
+		if (!mobj) {
+			/* if running Linux, check the value of the kernel
+			 * config CONFIG_OPTEE_SHM_NUM_PRIV_PAGES
+			 */
+			EMSG("not enough memory");
+			return TEE_ERROR_OUT_OF_MEMORY;
+		}
+
+		assert(mobj_get_va(mobj, 0));
+	}
+
```

Comment:
Need to add a mutex (OP-TEE condvar) to protect allocated buffer from multi-thread access.

Reply:
absolutely. I forgot I had disabled multithread support on op-tee for my work. will do.

Reply:
not anymore since the buffer is now thread specific

Reply:
fine, thanks.